### DISCUSSION TITLE
inquiry_complete_cb() callback function is never called when attempting to mount FAT12 file system

### DISCUSSION THREAD

**Discussion by Asherza** [hathach/tinyusb#2199] [Q&A]
Hello,

I am working on a project where I am attempting to mount an old camera that when plugged in acts as a mass storage device. When plugging into my computer I am able to tell what type of drive it is, as well as the format: FAT12

![image](https://github.com/hathach/tinyusb/assets/11355984/d6699a9f-bb8d-4353-9d06-8aaccfb524fc)

Now when I plug in the drive to my micro (using a pico usb drive connected via usb), It doesn't seem to call back my msc implementation. I know the implementation works because I am able plug in a USB stick and get the inquiry_complete_cb() to run without problem. 

The code base I am working out of is a forked [version](https://github.com/rppicomidi/pico-usb-host-msc-demo/tree/main) of the msc-file-explorer that is just set up to be configured for a pi pico.  

I set my CFG_TUSB_DEBUG to 3 and attempted follow along with the debug log, but I am very much so a beginner when it comes to using tusb. 

I have provided links to my debug logs:
[USB](https://pastebin.com/y6UwXBw0)
[Camera](https://pastebin.com/E0mcbNUs)

At the end of the USB debug log you can see the output of my inquiry_complete_cb

`SanDisk  U3 Cruzer Micro  rev 4.05
Disk Size: 1952 MB
Block Count = 3999373, Block Size: 512`

Please let me know if you need me to provide anymore information related to the problem. 
Again, I am quite new to the whole tusb stack, so I tried my best to provide information that I thought would be useful

Thank you in advance!

**Comment by Rocky04** [comment]
So check the expected stack trance to see when it derivate. I don't have any clue how the msc stuff should work.

```
[msc_host.c] msch_xfer_cb
[msc_host.c] config_get_maxlun_complete
([msc_host.c] config_request_sense_complete)
[msc_host.c] config_test_unit_ready_complete()
[msc_host.c] config_read_capacity_complete()
[msc_app] tuh_msc_mount_cb()
[msc_app] inquiry_complete_cb()
```

**Reply by Asherza** [reply]
Hey Rocky, is there a great way to dump the stack trace? I could go through and do some debug print statements, but if there is a better way, i'd like to use that. 

Thank you!

**Reply by Rocky04** [reply]
I manually checked it by tracing back who is calling the missed `inquiry_complete_cb()`. It's also not that you have an issue and want to figure out who was calling it but you would be interested the other way around to figure out why the function isn't called.

So no, I don't think there is a better way. You can either try to start with a deeper function more close to the one you are interested in to see if these are called until you find the last one which is, to check why it's not going deeper or you can try with the highest and then step through to see if the next deeper one is entered. But the later may provide some hits which you aren't interested in and it can be that it breaks due to the pause in execution.

_Edit:_
As mentioned I'm not familiar with the code so I can't say where potential issues are. Assuming the code is fine the problem is maybe an issue with the first callback and / or the code of the ported MCU.

**Comment by Asherza** [comment]
I had to put this down for a bit, so I haven't responded to this tread for a while. I recently picked this back up and started digging into some of the source code a bit. It seems I am getting a 
![image](https://github.com/hathach/tinyusb/assets/11355984/48e473a1-58f8-4907-9283-d1826db21559)
```protocol is not supported``` report when I plug in my camera. It seems I am not getting this error. It seems like in usbh.c on line 1677 the code is trying to find a driver for the interface, and is unable and spits out this error. 

More specifically, it seems to struggle doing the driver->open() on line 1681;
![image](https://github.com/hathach/tinyusb/assets/11355984/c11edda5-3656-4e0c-8601-d8e1cd3716ee)

When jumping into the open(), it seems to fail at line 393 on the TU_VERIFY
![image](https://github.com/hathach/tinyusb/assets/11355984/6195d910-e634-4f6f-9031-15a970611902)


Due to my unfamiliarity of the tinyUSB system, does anyone have any idea on where to go from here?

Thank you in advance!

**Reply by Rocky04** [reply]
The TinyUSB MSC host only seems to support the "SCSI transparent command set" (`MSC_SUBCLASS_SCSI`) but the attached device seems to use an implementation "Specific to device vendor" (`bInterfaceSubClass == 0xff`).
See [Mass Storage Class Specification Overview 1.4](https://usb.org/sites/default/files/Mass_Storage_Specification_Overview_v1.4_2-19-2010.pdf).

Meaning the device you want to attach seems to need a custom driver in order to operate and you would need to implement that into TinyUSB in order to get it to work.

_Edit:_
So you would need to check how the camera is working and sending data. Because it's vendor defined you would need to write a custom MSC driver just for this device by filtering the vendor and product ID. Then you would need to inject that via `usbh_app_driver_get_cb`.

_Edit 2:_
A modern OS has a collection of common custom drivers provided by the vendors. Like for an older system you may need to install a driver manually while you may don't need that anymore on a newer one. Some more rare ones are maybe downloaded from a server if needed. It may helps to check your OS to see which driver is applied to the device nodes of the camera.

A typically device tree looks like this:
(USB composite device) -> USB Devices -> functions

For a USB stick this can be like this:
USB Mass Storage Device [Win USB storage driver "USBSTOR"] -> USB Flash Storage Device [the actual drive / disk via the Win disk driver "DISK"] -> Volume [a partition of the device via a file system driver]
Here there is only a single USB device and so it's not a composite device. The function would be the storage device, so that is decoupled from the USB device while related to it. The volume on the other side has nothing to anymore with the USB device and so only relate to the specific disk.

So on Win you can check if there is a custom driver provided from the vendor applied to the USB device node or if the generic USBSTOR driver is used. If the later is used it can be that the implementation don't derivate from the standard that much. Or better, you can check if the device is registered under Linux and then check which driver is used and so check out the source code of that.

Think of it that the USB device is like a device controller. There can be a SCSI, IDE or SATA controller and the host needs to support the specific type. For that it doesn't matter if there is a driver how to read and write to a block device and how to handle different file systems.
